Automattic/simplenote-macos
# Extract all build settings to xcconfigs

For my 2020 Q4 HACK week project, I decided to look into extracting all of the project configurations into `xcconfig` files.

## Rationale

Having the build settings in dedicated files brings a few advantages:

- When the configs change, the resulting diff is cleaner and easier to review
- We can add comments in the `xcconfig`s, which helps adding context to configurations that might not be standard
- Extracting configurations also helps in case we'll move towards project generators like [Tuist](https://tuist.io/) or better build systems like [Bazel](https://bazel.build/) to help us better manage the project's complexity and build performance.

## The Process

I started with Simplenote macOS because it's the one with the simplest build configurations and target setup.

To extract the configs, I used [BuildSettingsExtractor app](https://buildsettingextractor.com/) made by @dempseyatgithub. The app uses a neat structure for the `xcconfig`s it generates: the project and all of the targets have each a shared `xcconfig` which is imported by the build configuration files.

The app also adds helpful documentation comments for each setting on which they are available. These comments, while helpful, also add a lot of lines to the diff. I'm conflicted between keeping them or not. I opened #709 to show this diff but without comments, keen to know what you think of it. 

### DRYing the setup

I attempted to DRY the configs setup a bit by extracting a couple of common settings I noticed while reading the files: the default code signing and the Swift optimization setting for the Debug builds.

I'm sure there are other opportunities to extra configurations in composable units, but I time-boxed my experiment.

## Future proofing

To avoid new settings making it into the project and being missed during code review, I added a Run Script build phase that runs the [`VerifyNoBS`](https://github.com/olofhellman/VerifyNoBS) script by @olofhellman. This script parses the project file and fails the build if there are build settings.

## To Test

This PR touches all the targets and all the build configurations. I'm not sure how to test it thoroughly, but here's what I did:

- Run the unit tests
- Archive both schemes, and validated the App Store one with ASC via the Xcode UI
- Made sure there are no new warnings

CI also builds and tests the Simplenote scheme and as you can see it's ✅ 

To test that the build fails when adding a new build settings to the project, I added a new build setting to the project and tried to build 😄 

<img width="1443" alt="Screen Shot 2020-11-06 at 12 26 05 pm" src="https://user-images.githubusercontent.com/1218433/98315052-5fed4580-202b-11eb-98d9-c81a540aa660.png">

## Feedback Please

I'm keen to know what you think of this approach. I think it's worth pursuing because of the benefit listed above.

The many config files setup seems cumbersome at first, but I believe it's granularity and consistency will make it easier to maintain in the long run. I also feel that if we find more opportunities to DRY the leaf config files, it will become easier to wrap our head around them.

Other questions:

- Is there anything I might have missed in the setup or in testing it that might bite us later down the track?
- What about the documentation in the files, keep the comments? No comments? _Some_ comments?

Thanks 😄
